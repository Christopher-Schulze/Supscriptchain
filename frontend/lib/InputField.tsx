'use client';
import * as Form from '@radix-ui/react-form';

interface InputFieldProps {
  name: string;
  label: string;
  value: string;
  onChange: (v: string) => void;
  type?: string;
  error?: string | null;
}

export default function InputField({
  name,
  label,
  value,
  onChange,
  type = 'text',
  error,
}: InputFieldProps) {
  return (
    <Form.Field name={name} className="field">
      <Form.Label htmlFor={name}>{label}</Form.Label>
      <Form.Control asChild>
        <input id={name} value={value} onChange={(e) => onChange(e.target.value)} type={type} />
      </Form.Control>
      {error && <Form.Message className="error">{error}</Form.Message>}
    </Form.Field>
  );
}
