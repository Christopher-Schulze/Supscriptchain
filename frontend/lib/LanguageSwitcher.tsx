'use client';
import { useTranslation } from 'react-i18next';

export default function LanguageSwitcher() {
  const { i18n } = useTranslation();
  const toggle = () => {
    i18n.changeLanguage(i18n.language === 'en' ? 'de' : 'en');
  };
  return (
    <button onClick={toggle} className="language-switcher">
      {i18n.language === 'en' ? 'DE' : 'EN'}
    </button>
  );
}
