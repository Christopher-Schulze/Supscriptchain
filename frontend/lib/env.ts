import { z } from 'zod';

const envSchema = z.object({
  NEXT_PUBLIC_CONTRACT_ADDRESS: z.string().nonempty(),
  NEXT_PUBLIC_CHAIN_ID: z.coerce.number(),
  NEXT_PUBLIC_RPC_URL: z.string().url(),
  NEXT_PUBLIC_SUBGRAPH_URL: z.string().url(),
});

export const env = envSchema.parse({
  NEXT_PUBLIC_CONTRACT_ADDRESS: process.env.NEXT_PUBLIC_CONTRACT_ADDRESS,
  NEXT_PUBLIC_CHAIN_ID: process.env.NEXT_PUBLIC_CHAIN_ID,
  NEXT_PUBLIC_RPC_URL: process.env.NEXT_PUBLIC_RPC_URL,
  NEXT_PUBLIC_SUBGRAPH_URL: process.env.NEXT_PUBLIC_SUBGRAPH_URL,
});

export type Env = typeof env;
