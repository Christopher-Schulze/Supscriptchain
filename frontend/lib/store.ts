'use client';
import { createContext, useContext, useState } from 'react';

interface State {
  message: string | null;
  setMessage: (m: string | null) => void;
}

const Ctx = createContext<State | undefined>(undefined);

export function StoreProvider({ children }: { children: React.ReactNode }) {
  const [message, setMessage] = useState<string | null>(null);
  return <Ctx.Provider value={{ message, setMessage }}>{children}</Ctx.Provider>;
}

export function useStore() {
  const ctx = useContext(Ctx);
  if (!ctx) throw new Error('StoreProvider missing');
  return ctx;
}
