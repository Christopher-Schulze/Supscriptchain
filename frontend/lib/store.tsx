'use client';
import { createContext, useContext, useState } from 'react';

export interface Message {
  text: string;
  type?: 'success' | 'warning' | 'error';
}

interface State {
  message: Message | null;
  setMessage: (m: Message | null) => void;
}

const Ctx = createContext<State | undefined>(undefined);

export function StoreProvider({ children }: { children: React.ReactNode }) {
  const [message, setMessage] = useState<Message | null>(null);
  return <Ctx.Provider value={{ message, setMessage }}>{children}</Ctx.Provider>;
}

export function useStore() {
  const ctx = useContext(Ctx);
  if (!ctx) throw new Error('StoreProvider missing');
  return ctx;
}
