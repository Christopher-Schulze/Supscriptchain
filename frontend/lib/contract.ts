import { ethers } from "ethers";
import type { ExternalProvider } from "ethers";
import type { Subscription } from "typechain/contracts/Subscription.sol/Subscription";
import { Subscription__factory } from "typechain/factories/contracts/Subscription.sol/Subscription__factory";
import { env } from "./env";

function parseEthersError(err: unknown): string {
  const e = err as any;
  return (
    e?.shortMessage ||
    e?.reason ||
    e?.error?.message ||
    e?.data?.message ||
    (err instanceof Error ? err.message : String(err))
  );
}

async function handleEthersError<T>(fn: () => Promise<T>): Promise<T> {
  try {
    return await fn();
  } catch (err) {
    throw new Error(parseEthersError(err));
  }
}

declare global {
  interface Window {
    ethereum?: ExternalProvider;
  }
}

export async function getContract(): Promise<Subscription> {
  const address = env.NEXT_PUBLIC_CONTRACT_ADDRESS;
  let signerOrProvider: ethers.Signer | ethers.Provider;
  if (typeof window !== "undefined" && (window as Window).ethereum) {
    const provider = new ethers.BrowserProvider((window as Window).ethereum!);
    const signer = await provider.getSigner();
    signerOrProvider = signer;
  } else {
    const rpc = env.NEXT_PUBLIC_RPC_URL;
    signerOrProvider = new ethers.JsonRpcProvider(rpc);
  }
  return Subscription__factory.connect(address, signerOrProvider);
}

export async function subscribeWithPermit(
  planId: bigint,
  deadline: bigint,
  v: number,
  r: string,
  s: string
) {
  const contract = await getContract();
  return handleEthersError(() =>
    contract.subscribeWithPermit(planId, deadline, v, r, s)
  );
}

export async function createPlan(
  merchant: string,
  token: string,
  price: bigint,
  billing: bigint,
  priceInUsd: boolean,
  usdPrice: bigint,
  feed: string
) {
  const contract = await getContract();
  return handleEthersError(() =>
    contract.createPlan(
      merchant,
      token,
      price,
      billing,
      priceInUsd,
      usdPrice,
      feed
    )
  );
}

export async function updatePlan(
  planId: bigint,
  billing: bigint,
  price: bigint,
  priceInUsd: boolean,
  usdPrice: bigint,
  feed: string
) {
  const contract = await getContract();
  return handleEthersError(() =>
    contract.updatePlan(planId, billing, price, priceInUsd, usdPrice, feed)
  );
}

export async function updateMerchant(planId: bigint, merchant: string) {
  const contract = await getContract();
  return handleEthersError(() => contract.updateMerchant(planId, merchant));
}
