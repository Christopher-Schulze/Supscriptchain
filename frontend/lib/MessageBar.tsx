'use client';
import { useStore } from './store';
import { useTranslation } from 'react-i18next';

export default function MessageBar() {
  const { message, setMessage } = useStore();
  const { t } = useTranslation();
  if (!message) return null;
  const className = `message-bar ${message.type ?? ''}`.trim();
  return (
    <div
      className={className}
      role="button"
      aria-live="assertive"
      tabIndex={0}
      aria-label={t('messages.dismiss')}
      onClick={() => setMessage(null)}
      onKeyDown={(e) => {
        if (e.key === 'Enter' || e.key === ' ') {
          setMessage(null);
        }
      }}
    >
      {message.text}
    </div>
  );
}
