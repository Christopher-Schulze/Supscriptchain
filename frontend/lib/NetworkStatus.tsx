'use client';
import { useEffect, useState } from 'react';
import { ethers } from 'ethers';
import { env } from './env';
import { useStore } from './store';

export default function NetworkStatus() {
  const [name, setName] = useState('');
  const [block, setBlock] = useState<number | null>(null);
  const { setMessage } = useStore();

  useEffect(() => {
    let cancelled = false;
    const provider: ethers.Provider =
      typeof window !== 'undefined' && (window as any).ethereum
        ? new ethers.BrowserProvider((window as any).ethereum)
        : new ethers.JsonRpcProvider(env.NEXT_PUBLIC_RPC_URL);

    async function load() {
      try {
        const net = await provider.getNetwork();
        const blk = await provider.getBlockNumber();
        if (!cancelled) {
          setName(net.name === 'unknown' ? String(net.chainId) : net.name);
          setBlock(blk);
        }
      } catch (err) {
        if (!cancelled) {
          const msg = err instanceof Error ? err.message : 'Failed to connect';
          setMessage({ text: msg, type: 'error' });
        }
      }
    }

    load();
    const id = setInterval(load, 10000);
    return () => {
      cancelled = true;
      clearInterval(id);
    };
  }, [setMessage]);

  if (!name || block === null) return null;

  return (
    <span className="network-status" style={{ float: 'right', marginLeft: 10 }}>
      {name} #{block}
    </span>
  );
}
