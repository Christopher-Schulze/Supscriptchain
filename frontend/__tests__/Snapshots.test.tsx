import { render } from '@testing-library/react';
import Manage from '../app/manage/page';
import Payment from '../app/payment/page';
import Plans from '../app/plans/page';
import { StoreProvider } from '../lib/store';

jest.mock('../lib/useWallet', () => () => ({ account: '0xabc', connect: jest.fn() }));
jest.mock('../lib/plansStore', () => ({ usePlans: () => ({ plans: [], reload: jest.fn() }) }));
jest.mock('../lib/useUserSubscriptions', () => () => ({ subs: [], reload: jest.fn() }));

function Wrapper({ children }: { children: React.ReactNode }) {
  return <StoreProvider>{children}</StoreProvider>;
}

test('manage page snapshot', () => {
  const { asFragment } = render(
    <Wrapper>
      <Manage />
    </Wrapper>
  );
  expect(asFragment()).toMatchSnapshot();
});

test('plans page snapshot', () => {
  const { asFragment } = render(
    <Wrapper>
      <Plans />
    </Wrapper>
  );
  expect(asFragment()).toMatchSnapshot();
});

test('payment page snapshot', () => {
  const { asFragment } = render(
    <Wrapper>
      <Payment />
    </Wrapper>
  );
  expect(asFragment()).toMatchSnapshot();
});
