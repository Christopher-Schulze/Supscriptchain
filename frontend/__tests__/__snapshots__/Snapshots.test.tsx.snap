// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manage page snapshot 1`] = `
<DocumentFragment>
  <div
    aria-busy="false"
  >
    <h1>
      Manage Subscription
    </h1>
    <form
      class="form"
    >
      <div
        class="field"
      >
        <label
          for="plan-id"
        >
          Plan ID:
        </label>
        <input
          id="plan-id"
          name="plan-id"
          title=""
          type="text"
          value="0"
        />
      </div>
      <div
        class="field"
      >
        <label
          for="deadline"
        >
          Deadline (unix secs):
        </label>
        <input
          id="deadline"
          name="deadline"
          title=""
          type="text"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <label
          for="sig-v"
        >
          v:
        </label>
        <input
          id="sig-v"
          name="sig-v"
          title=""
          type="text"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <label
          for="sig-r"
        >
          r:
        </label>
        <input
          id="sig-r"
          name="sig-r"
          title=""
          type="text"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <label
          for="sig-s"
        >
          s:
        </label>
        <input
          id="sig-s"
          name="sig-s"
          title=""
          type="text"
          value=""
        />
      </div>
      <div
        style="display: flex; gap: 8px; flex-wrap: wrap;"
      >
        <button
          type="button"
        >
          Get Permit Signature
        </button>
        <button
          type="button"
        >
          Subscribe with Permit
        </button>
        <button
          type="submit"
        >
          Subscribe
        </button>
        <button
          type="button"
        >
          Cancel
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`payment page snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Process Payment
    </h1>
    <form
      class="form"
    >
      <div
        class="field"
      >
        <label
          for="pay-user"
        >
          User:
        </label>
        <input
          id="pay-user"
          name="pay-user"
          title=""
          type="text"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <label
          for="pay-plan"
        >
          Plan ID:
        </label>
        <input
          id="pay-plan"
          name="pay-plan"
          title=""
          type="text"
          value="0"
        />
      </div>
      <button
        type="submit"
      >
        Process
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`plans page snapshot 1`] = `
<DocumentFragment>
  <div
    aria-busy="false"
  >
    <h1>
      Available Plans
    </h1>
    <div
      style="margin-bottom: 10px;"
    >
      <a
        href="/plans/create"
      >
        Create Plan
      </a>
       | 
      <a
        href="/plans/manage"
      >
        Manage Plans
      </a>
    </div>
    <form
      style="display: flex; align-items: center; gap: 8px; flex-wrap: wrap; margin-bottom: 10px;"
    >
      <div
        class="field"
      >
        <label
          for="filter"
        >
          Filter
        </label>
        <select
          id="filter"
          name="filter"
          title=""
        >
          <option
            value="all"
          >
            All
          </option>
          <option
            value="active"
          >
            Active
          </option>
          <option
            value="inactive"
          >
            Inactive
          </option>
        </select>
      </div>
      <div
        class="field"
      >
        <label
          for="sort"
        >
          Sort
        </label>
        <select
          id="sort"
          name="sort"
          title=""
        >
          <option
            value="asc"
          >
            Price ascending
          </option>
          <option
            value="desc"
          >
            Price descending
          </option>
        </select>
      </div>
    </form>
    <ul
      class="list"
    />
  </div>
</DocumentFragment>
`;
